### ISSUE TITLE
rpc: dumptxoutset as sqlite file

### ISSUE THREAD

**Issue by prusnak** [bitcoin/bitcoin#24628]
There has been at least 3 efforts to expand the `dumptxoutset` command to be able to export the data as CSV: 
* https://github.com/bitcoin/bitcoin/pull/18689
* https://github.com/bitcoin/bitcoin/pull/24202
* [luke-jr/bitcoin#rpc_dumptxoutset_hr](https://github.com/luke-jr/bitcoin/tree/rpc_dumptxoutset_hr)

JSON export also has been requested: https://github.com/bitcoin/bitcoin/issues/21670

Lately [@theStack suggested](https://github.com/bitcoin/bitcoin/pull/24202#issuecomment-1064124278) that we can leverage the SQLite format for the export.

I think this is the best approach since it is trivial to export both CSV and JSON from SQLite + you can access the file programatically via virtually any programming language:
* `sqlite3 -header -csv foo.db "select * from foo" > foo.csv`
* `sqlite3 -json foo.db "select * from foo" > foo.json`

... and we already have SQLite as an optional dependency. Such format is also future proof (we can add new table columns if needed).

**Comment by jamesob** [comment]
Concept ACK; happy to review any related PRs.

**Comment by dunxen** [comment]
Sounds good! I can take a look at this.

**Comment by dunxen** [comment]
So I should have a PR up for this by the end of the week.

**Comment by theStack** [comment]
Some thoughts that I wanted to share for discussion: a pretty simple alternative to directly support dumping in SQLite format in bitcoind would be to provide a small external tool (e.g. in `contrib`, or even outside the repository) that simply takes the compact-serialized UTXO set format as it created by `dumptxoutset` now as input and create the SQLite database out of that. The drawback of that approach is of course that it's a two-step process that hence also takes more time and space (the latter at least temporarily), but on the other hand the user could more easily adapt the details of the created database (which columns do we even want? for what columns do we want indices? what metadata should be included in the database? etc.) without the need to recompile bitcoind or wait for another release. There could be lots of potential PRs coming that will try to bikeshed or add extra data to the output, like address strings, timestamps of the UTXO creation etc... where in the end the use-case and trade-off between time/space and amount of needed information per UTXO is highly individual. For example, many users probably only want the outputs (scriptPubKey and nAmount) to see "who owns what" and don't care about the outpoints (prevoutHash and prevoutIndex).

Such a `utxos_converter` tool could be made in a generic way that is easily extendable in a way that pretty much any output format can be provided, by having a interface that takes an `init` and a `write_utxo_entry` function which can be specified by its own needs.

Thoughts?

**Comment by sipa** [comment]
> Nevermind, I just found out that for decompressing a P2PK scriptPubKey one needs secp256k1

Why would a dump tool need to do anything with public keys except pass them through?

**Comment by theStack** [comment]
> > Nevermind, I just found out that for decompressing a P2PK scriptPubKey one needs secp256k1
> 
> Why would a dump tool need to do anything with public keys except pass them through?

See the `DecompressFunction` for special types 4 and 5 (that is, P2PK with uncompressed pubkeys). The pubkey is stored as compressed in the compact-serialized UTXO set and then uncompressed via secp256k1:

https://github.com/bitcoin/bitcoin/blob/0342ae1d395ca82614f6d3b8fabb6a44403baf2a/src/compressor.cpp#L123-L135

**Comment by sipa** [comment]
Oh, I didn't realize the current dump format uses the UTXO database script compression. That seems undesirable in any case, TBH, independent from the question of sqlite or anything else.

**Comment by theStack** [comment]
> Oh, I didn't realize the current dump format uses the UTXO database script compression. That seems undesirable in any case, TBH, independent from the question of sqlite or anything else.

Agreed, the current dump format isn't very helpful (at least not from a user's perspective).

FWIW, I've written a tool outlined in https://github.com/bitcoin/bitcoin/issues/24628#issuecomment-1108469715 that converts the compact-serialized format to a SQLite database, ~200 LOC in golang: https://github.com/theStack/utxo_to_sqlite
Most of the code is decompression routines taken from Bitcoin Core (VARINTs, amounts and scriptPubKeys). It currently skips P2PK outputs with uncompressed pubkeys (for the reasons stated above) and hence can only provide an incomplete UTXO set, but seems to work fine otherwise. After #24952 is merged, there will be no need for this anymore, but maybe someone finds this useful until then or interesting to play around with (I saw it as a welcome opportunity to get a bit more familiar with golang).

**Comment by fjahr** [comment]
I think this is resolved with the merged of #27432?

**Comment by prusnak** [comment]
> I think this is resolved with the merged of [#27432](https://github.com/bitcoin/bitcoin/pull/27432)?

Yes
